### PR TITLE
Allow human players to disconnect without stopping game

### PR DESCRIPTION
**Pull request by o01eg**
Next step of #2243 

Adds two options. The first is to define minimal connected human player to don't stop playing game. The second is to define maximum disconnected human player to don't stop playing game.
Default values acts as current setting when server stops game if no human player left or any alive empire was disconnected.

### CODE REVIEW

File: default/stringtables/en.txt [hunk 1]
```diff
@@ -1384,6 +1384,12 @@ Set the minimum number of human players required to start a multiplayer game.
 OPTIONS_DB_MP_HUMAN_MAX
 Limit the number of human players allowed in a multiplayer game. Set to -1 for unlimited.
 
+OPTIONS_DB_MP_CONN_HUMAN_MIN
+Server stop the playing game if count of active human player will be less than this value.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Minimum number of active human players, below which the server will stop the game being played.

File: default/stringtables/en.txt [hunk 2]
```diff
@@ -1384,6 +1384,12 @@ Set the minimum number of human players required to start a multiplayer game.
...
+
+OPTIONS_DB_MP_DISCONN_HUMAN_MAX
+Server stop the playing game if count of disconnected human player from non-eliminated empires exceeds or equals this value. Disabled if 0.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Maximum number of players, who control non-eliminated empires, who may disconnect from the server before it stops the game being played. Disabled if 0, meaning the server will never stop the game due to empire-playing players disconnecting.

of active human players, below which the server will stop the game being played.

File: server/ServerApp.cpp [hunk 3]
```diff
@@ -1757,14 +1757,14 @@ int ServerApp::AddPlayerIntoGame(const PlayerConnectionPtr& player_connection) {
     }
 
     if (empire_id == ALL_EMPIRES)
-        return false;
```
Comment (geoffthemedio) [inline server/ServerApp.cpp:1767]: is this change relevant to the subject of the pull request?

Comment (o01eg) [reply server/ServerApp.cpp:1767]: Reconnection doesn't work properly without this fix.

File: default/stringtables/en.txt (hunk 1 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Fixed.

File: default/stringtables/en.txt (hunk 2 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Fixed.

File: default/stringtables/en.txt [hunk 4]
```diff
@@ -1384,6 +1384,12 @@ Set the minimum number of human players required to start a multiplayer game.
 OPTIONS_DB_MP_HUMAN_MAX
 Limit the number of human players allowed in a multiplayer game. Set to -1 for unlimited.
 
+OPTIONS_DB_MP_CONN_HUMAN_MIN
+Minimum number of active human players, below which the server will stop the game being played.
+
+OPTIONS_DB_MP_DISCONN_HUMAN_MAX
+Maximum number of players, who control non-eliminated empires, who may disconnect from the server before it stops the game being played. Disabled if 0, meaning the server will never stop the game due to empire-playing players disconnecting.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Is this actually the number of times a player can disconnect, ignoring any reconnect, or the number of empires that must have disconnected players, at the same time, for the server to abort?

eg. if a player disconnects, reconnects, disconnects, and reconnects, is that 2 or 0 disconnects / disconnected players?

File: server/ServerFSM.cpp [hunk 5]
```diff
@@ -241,50 +241,79 @@ ServerApp& ServerFSM::Server()
...
             }
         }
     } else {
         DebugLogger(FSM) << "Client quit before id was assigned.";
     }
 
-    // independently of everything else, if there are no humans left, it's time to terminate
-    if (m_server.m_networking.empty()
-        || m_server.m_ai_client_processes.size() == m_server.m_networking.NumEstablishedPlayers())
+    int connected_count = 0;
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: add a comment here explaining what is being tallied

File: server/ServerFSM.cpp [hunk 6]
```diff
@@ -241,50 +241,79 @@ ServerApp& ServerFSM::Server()
...
+                break;
+            default:
+                ErrorLogger(FSM) << "Incorrect client type " << m_server.GetEmpireClientType(empire.first)
+                                 << " for empire #" << empire.first;
+                break;
+            }
+        }
+    }
+
+    // Stop server if connected human player count is less than limit
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: limit -> minimum ?

File: server/ServerFSM.cpp (hunk 5 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp]: also, might this be better described as a count of the empires that should have human players, which have or don't currently have a connected or disconnected human player?

File: default/stringtables/en.txt (hunk 4 above)
Comment (o01eg) [reply default/stringtables/en.txt]: It's a number of empires that have disconnected players, at the same time.

File: server/ServerFSM.cpp (hunk 5 above)
Comment (o01eg) [reply server/ServerFSM.cpp]: It's a count of active (non-eliminated) empires, which currently have a connected human players. Eliminated or AI players don't count.

File: default/stringtables/en.txt (hunk 4 above)
Comment (geoffthemedio) [reply default/stringtables/en.txt]: so then `disconnect` -> `be disconnected` and `disconnecting` -> `being disconnected`

File: server/ServerFSM.cpp (hunk 5 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp]: I'd give these variables clearer names, then

File: server/ServerFSM.cpp (hunk 6 above)
Comment (o01eg) [reply server/ServerFSM.cpp]: Fixed.

File: server/ServerFSM.cpp (hunk 5 above)
Comment (o01eg) [reply server/ServerFSM.cpp]: Changed names and added comments.

File: default/stringtables/en.txt (hunk 4 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Fixed.

**Comment by o01eg** [comment]
Fixed collision. Rebased.

**Comment by o01eg** [comment]
Is something left to fix to accept PR?
